src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonPropertyInfo.cs

```diff
@@ -341,269 +172,77 @@ public virtual void GetPolicies()
 
         // The name of the property with any casing policy or the name specified from JsonPropertyNameAttribute.
         public byte[]? Name { get; private set; }
-        public string? NameAsString { get; private set; }
+        public string? NameAsString { get; set; }
 
         // Key for fast property name lookup.
         public ulong PropertyNameKey { get; set; }
 
         // Options can be referenced here since all JsonPropertyInfos originate from a JsonClassInfo that is cached on JsonSerializerOptions.
         protected JsonSerializerOptions Options { get; set; } = null!; // initialized in Init method
 
-        protected abstract void OnRead(ref ReadStack state, ref Utf8JsonReader reader);
-        protected abstract void OnReadEnumerable(ref ReadStack state, ref Utf8JsonReader reader);
-        protected abstract void OnWrite(ref WriteStackFrame current, Utf8JsonWriter writer);
-        protected virtual void OnWriteDictionary(ref WriteStackFrame current, Utf8JsonWriter writer) { }
-        protected abstract void OnWriteEnumerable(ref WriteStackFrame current, Utf8JsonWriter writer);
-
-        public Type ParentClassType { get; private set; } = null!;
-
-        public PropertyInfo? PropertyInfo { get; private set; }
-
-        public void Read(JsonTokenType tokenType, ref ReadStack state, ref Utf8JsonReader reader)
+        public bool ReadJsonAndAddExtensionProperty(object obj, ref ReadStack state, ref Utf8JsonReader reader)
         {
-            Debug.Assert(ShouldDeserialize);
+            object propValue = GetValueAsObject(obj)!;
```

Comment:
The extension data property value is created in `CreateDataExtensionProperty`. I don't believe this PR changed any logic around that.

Reply:
Sure, that makes sense. If `propValue` is guaranteed to not be null, and we check elsewhere that `JsonExtensionData` attribute can only be put on `IDictionary<string, object>` or `JsonElement`, then why do we **try** to cast it and check for null below rather than doing an explicit cast and removing the checks?

Will `dictionaryObject` ever be `null` in this condition:
```C#
if (dictionaryObject != null && reader.TokenType == JsonTokenType.Null)
{
    // A null JSON value is treated as a null object reference.
    dictionaryObject[state.Current.JsonPropertyNameAsString!] = null;
}
```

The existence of other checks in this `ReadJsonAndAddExtensionProperty` method made me think `propValue` can be null, and hence why I thought maybe we should annotate it as such.

This isn't super important to guard against, but the following incorrectly implemented extension dictionary will cause a null ref. We have the same problem in master:
```C#
[Fact]
public static void QuickTest()
{
    JsonSerializer.Deserialize<TestClass>("{\"prop\": \"1\"}");
}
public class TestClass
{
    Dictionary<string, object> _dictionary = null;

    [JsonExtensionData]
    public Dictionary<string, object> Dictionary
    {
        get
        {
            return _dictionary;
        }
        set
        {
            if (_dictionary != null)
            {
                _dictionary = value;
            }
        }
    }
}
```